easy configuration for freemarker

jeffwilde opened the pull request:
This pull request externalizes the freemarker Configuration instantiation code into a true factory class. 

The new code structure provides the following benefits: 
- makes it a lot easier to get a handle on the freemarker Configuration class before it is used by jersey
- simplifies the FreeMarkerViewProcessor class by externalizing the configuration setup in an external class
- makes it easy to reuse the jersey-specific Configuration setup (by simply extending the FreemarkerDefaultConfigurationFactory class)
- makes jersey's freemarker integration code structurally more similar to the mustache integration code

This would resolve [JERSEY-2771](https://java.net/jira/browse/JERSEY-2771).

mgajdos commented:
Hi Jeff, 
thanks for your contribution. 

Before we can review your pull request, we need you to sign [Oracle Contributor Agreement](http://www.oracle.com/technetwork/community/oca-486395.html).

More information about contributing to Jersey: https://jersey.java.net/scm.html#/Submitting_Patches_and_Contribute_Code…

Regards,
Michal

mgajdos commented:
Jenkins, please test this patch.

jeffwilde commented:
Michal, I emailed through the OCA yesterday, so I guess we're just waiting on whatever process Oracle has for those sorts of things.

shamoh commented:
Dear Universe, add to whitelist Mr. Jeff Wilde.

jeffwilde commented:
I actually have another couple commits I'd like to add to this before review.

jeffwilde commented:
Ok, I've added the extra commit.

I did not take into account backwards compatibility in my previous commits. Now I've added support for existing usage as a fallback to the new factory-style usage in my original PR.

Adding the feature didn't end up quite as clean as I had envisioned it, but I think it's still worth on your end.

jeffwilde commented:
Not sure if I can do anything about the build failure. I can't look any further into the integration test that actually failed (`jersey-tests-integration-servlet-3-chunked-io`), as the output from Jenkins just references the test run report file which is inaccessible from the web UI.

mgajdos commented:
@jeffwilde The failed test isn't your fault. We'll verify this change manually.

mgajdos commented:
Verified manually.

mgajdos commented on ext/mvc-freemarker/src/main/java/org/glassfish/jersey/server/mvc/freemarker/FreemarkerConfigurationFactory.java:
```diff
@@ -0,0 +1,48 @@
+
+import freemarker.template.Configuration;
+
+public interface FreemarkerConfigurationFactory {
```
Missing Javadocs.

mgajdos commented on ext/mvc-freemarker/src/main/java/org/glassfish/jersey/server/mvc/freemarker/FreemarkerDefaultConfigurationFactory.java:
```diff
@@ -0,0 +1,95 @@
+import java.io.IOException;
+import java.util.List;
+
+import freemarker.cache.*;
```
No .* imports.

mgajdos commented on ext/mvc-freemarker/src/main/java/org/glassfish/jersey/server/mvc/freemarker/FreemarkerSuppliedConfigurationFactory.java:
```diff
@@ -0,0 +1,59 @@
+
+import freemarker.template.Configuration;
+
+public class FreemarkerSuppliedConfigurationFactory implements FreemarkerConfigurationFactory {
```
This can be package private and final. Please, add some Javadocs.

mgajdos commented on ext/mvc-freemarker/src/main/java/org/glassfish/jersey/server/mvc/freemarker/FreemarkerViewProcessor.java:
```diff
@@ -99,25 +89,22 @@ public FreemarkerViewProcessor(final javax.ws.rs.core.Configuration config, fina
         super(config, servletContext, "freemarker", "ftl");
 
-        this.factory = getTemplateObjectFactory(serviceLocator, Configuration.class, new Value<Configuration>() {
+        ;
```
Remove ;

mgajdos commented on ext/mvc-freemarker/src/main/java/org/glassfish/jersey/server/mvc/freemarker/FreemarkerViewProcessor.java:
```diff
@@ -99,25 +89,22 @@ public FreemarkerViewProcessor(final javax.ws.rs.core.Configuration config, fina
-                    loaders.add(new WebappTemplateLoader(servletContext));
+            public FreemarkerConfigurationFactory get() {
+                Configuration configuration = getTemplateObjectFactory(serviceLocator, Configuration.class, new Value<Configuration>() {
+                    @Override
```
Use Values.empty() here.

mgajdos commented:
In addition to my comments in code: Provide some tests to cover added interfaces/classes.

mgajdos commented:
Jenkins, test this patch.

mgajdos commented:
Verified manually.

mgajdos commented:
Thank you for contribution!

jeffwilde commented:
You're welcome @mgajdos!  I did have plans to add some tests, though, once I figured out the best place to put them. 

I didn't think unit-tests would be very helpful without a mocking framework, so I was poking around the example Freemarker application (```examples/freemarker-webapp```) to see if I could shoehorn usage of this configuration mechanism in there. Let me know if you think this is still worthwhile or even the right dplace for the kind of testing you were thinking of.

mgajdos commented:
@jeffwilde Yes, it's still worthwhile - the example is a good place to show this new way of providing Configuration for Freemarker. Thanks!
